### ISSUE TITLE
Common errors triggering assert `debugger;`

### ISSUE THREAD

**Issue by sjmiles** [PolymerLabs/arcs#2149]
Background: for web-dev, the `debugger` command in `assert-web.js` is an annoyance when it is triggered frequently. IIRC, we agreed to leave it in, but only use it for truly exceptional situations, and switch more common problems to use proper exceptions.

There are some errors I contend with repeatedly (enough to make me want to fork `assert-web.js` and remove the `debugger`, which is a hassle to avoid checking in accidentally).

The first one is

Error: Cannot instantiate an unresolved recipe: recipe ...

P.S. I don't know to whom this ticket should be assigned, so I did my standard thing and assigned everybody, which is an attempt to ensure somebody sees it. Please let me know if this is annoying.

**Comment by sjmiles** [comment]
More info: this errors crop up a lot while developing recipes. Because of the multiplicity of workers, it's hard to mark it to 'ignore'. See what happens below, note the number of `assert-web` tabs in debugger.

![screen shot 2018-10-25 at 1 41 22 pm](https://user-images.githubusercontent.com/98045/47529226-11ade500-d85c-11e8-8843-49f9845a7bc2.png)

**Comment by shans** [comment]
Thanks Scott,

This is actually a good example of something that we should be fixing so I'm glad you've raised the issue.

The assert in question is in arc.instantiate, and it absolutely should be a truly exceptional situation, because it means something thought that it would be able to instantiate a recipe, but the recipe isn't capable of being instantiated. At this level (i.e. this place in the code) there's not really much we can do except assert.

I suspect that this is happening somewhere in speculative execution, otherwise there wouldn't be so many of these (and the effect even if ignored would be broken arcs). However, this is concerning because the strategizer machinery isn't ever supposed to provide unresolved recipes as outputs.

I'm doing a bit of investigation as to where this is coming from now. What would be the best way to repro?

**Comment by sjmiles** [comment]
For the record, I agree this situation is _exceptional_ and I don't have a problem with it throwing an exception, it's specifically that `debugger` statement which is problematic. As we discussed before, my sense was that we agreed to leave it in only for extra-exceptional situations.

Having said that, I'm hitting this because when I develop recipes today I skip the strategizer and go direct. This is a relatively new idea, but it's useful because it's still pretty hard to get a complex recipe right (at least for me).

**Comment by sjmiles** [comment]
I get this highly-useful stuff on console when this error occurs, this is great, but it's only after I deal with the debugger break. Don't mean to beat a dead horse, just trying to illuminate my perspective, sometimes I leave that bit out.

![screen shot 2018-10-25 at 2 06 39 pm](https://user-images.githubusercontent.com/98045/47530369-3d7e9a00-d85f-11e8-9e48-a47f2f37dbec.png)

**Comment by shans** [comment]
Right, I'm claiming that trying to instantiate unresolved recipes is extra-exceptional.

Which, it turns out is probably true because you're trying something completely new.

'Going direct' is difficult because you need to have a fully resolved recipe (all the handles pointing at real ids, all the slots hooked up, all the verbs resolved, etc. etc.) before the arc can understand it. We have a thing called a 'RecipeResolver' that can do this for you when there's only 1 candidate for each potential decision - it looks like you have one of these in arcs-env. Are you using it on the recipes before passing them into arc.instantiate?

**Comment by sjmiles** [comment]
> extra-exceptional.

Except it's happening to me dozens of times a day. That debugger statement has always been a stumbling block for me, I never really understood it's value, but IIRC we compromised on saying we would do something else for errors that I encounter a lot.

> Are you using it on the recipes before passing them into arc.instantiate?

Yes.

> 'Going direct' is difficult because ...

There are scenarios where it's useful to implement recipes directly. This is true for the Launcher arcs and the Pipe arcs and I assume other tooling the Shell will want to implement via arcs.

There is also the ongoing tension between monolithic recipes (that we tend to need at demo time) and more granular recipes (that benefit from strategizer/coalescer). I haven't found a way to have both, so I'm reluctant to start working on the latter again unless there is team buy-in; otherwise I'll get recipe whiplash like I did in October. I mention this because the former are easy to test without Strategizer and push me towards going direct.

**Comment by sjmiles** [comment]
Sorry if I'm putting too-fine a point on it. At some level it's just a practical problem for me, and I _can_ get around it by forking `assert-web.js`.

**Comment by shans** [comment]
My point is that if you're suddenly getting dozens of these then something systemic is going badly wrong.

In particular, if you're using RecipeResolver and the output recipes aren't resolved, that sounds like an egregious bug.

**Comment by shans** [comment]
Oh .. actually I can see what's going wrong. RecipeResolver doesn't return a recipe only if it's resolved; it returns one if it could *make some progress* towards resolution. We'll need to pull Maria in for a discussion of what the correct behaviour should be, but in the meantime you could follow up RecipeResolver invocation with a check for isResolved(), and console.log the recipe / abort at that point?

**Comment by sjmiles** [comment]
It's not a bug, it's that it's pretty hard to write recipes. It takes me numerous iterations to do any recipe work.

**Comment by sjmiles** [comment]
Ok, yes, I can work around it, but once more for the record, that `debugger` statement is never helpful for me, so I will inevitably end up removing it locally.

**Comment by shans** [comment]
I get that it's hard to write recipes.

But. Giving arc.instantiate() an unresolved recipe is a bad bug! At that point, we can't do anything with it except throw an error. Given the potential consequences (unsynced arcs, arcs that load for one user but not another, broken arcs, etc), I can't accept that this is anything but the most serious of errors, and hence I think it's entirely appropriate that there is a very squeaky wheel here.

If you want to create a class of errors that
(1) doesn't break into the debugger; but
(2) gets your attention strongly enough that you immediately report it in such a way that we drop everything and find out what's happening;
then maybe we can do that.

**Comment by sjmiles** [comment]
Fwiw, it's also possible that this comes up when something change that causes a serialization to stop being correct. This happens frequently as things evolve, but I will need to capture it live before I know if it's the same assert or not.

At the risk of repeating myself, this `debugger` statement is not so rare that it ever has the effect of #2. I hit these asserts all the time because of data-states and not because of any kind of code-errors. Also I find it generally strange to use `debugger` this way (note the eslint exception).

Anyway, I think we've gone enough rounds on it. I can just get around it locally (I understand you would prefer I not do that, but the alternatives are just to costly in the short-term).

**Comment by shans** [comment]
Let's discuss this in the team meeting? I don't want to get in your way here, but I'm also concerned that you're hitting these assertions at all (outside the context of this current one).

Like, they're all supposed to be things that mean the system can't guarantee consistency any more. If they're happening we should be finding out why and eliminating the causes (or the assertion, if we got it wrong, which can definitely happen too).

**Comment by sjmiles** [comment]
Recipes and serializations have dependencies that can shift as things develop. Changes in dependencies cause these low-level failures on a daily basis.

As I said, the other cause is me developing recipes and getting things wrong. 

I can trap these situations at a higher level, but I'm not sure that won't just mask them. Ironically, I used to test `isResolved()`, but I removed it because at the time it seemed redundant.
